SCons/Tool/gcc.py

```diff
@@ -74,14 +74,13 @@ def detect_version(env, cc):
     # GCC versions older than that, we should use --version and a
     # regular expression.
     # pipe = SCons.Action._subproc(env, SCons.Util.CLVar(cc) + ['-dumpversion'],
-    pipe=SCons.Action._subproc(env, SCons.Util.CLVar(cc) + ['--version'],
+    with SCons.Action._subproc(env, SCons.Util.CLVar(cc) + ['--version'],
```

Comment:
this shouldn't work - _subproc isn't a context manager (it's not a class with `__enter__` and `__exit__` methods) - I've made that mistake elsewhere, which I think somebody cleaned up.

Reply:
it returns the return value of `subprocess.Popen(cmd, **kw)`, which is a context handler right?

Reply:
No - the context manager protocol is applied to the thing the "with"  statement uses, not to the return. `with open(...) as f`: it applies to the open, not f, which it what open's `__enter__` method returned.

Reply:
[From docs..  ](https://docs.python.org/3/library/subprocess.html#popen-constructor)
```
Popen objects are supported as context managers via the with statement: on exit, standard file descriptors are closed, and the process is waited for.

with Popen(["ifconfig"], stdout=PIPE) as proc:
    log.write(proc.stdout.read())

```

Reply:
We need to revamp _subproc anyway (I have my thoughts, and there was the thread on it not too long ago).  However the existing change in the PR may be a useable band-aid in the meantime.  I applied the change and mypy didn't complain, which surprised me, because I've seen it complain on `with _subproc` somewhere else, i'm sure of it...

[on the revamp topic: _subproc has two purposes, one is to interpolate values from `env` if needed, the other is to keep the caller from having to do error handling - thus the return of the dummy-Popen object (which definitely isn't a context manager). The former is a valid use for the way SCons runs external processes, the latter could be handled in a better way.]

Reply:
So go ahead and merge then?

Reply:
yes, I was suggesting that as long as this quiets the complaints we're no worse off, and will address this chunk of the code eventually anyway.

Reply:
Sounds good.

Reply:
ugh.. 2 java tests failing with this change.

Reply:
I put this on Discord, but maybe better to have here.  The two Java tests whack the path to test something. That means when the env is initialized, the gcc check (which uses `_subproc`) fails since there's no such cmd now, and it returns a `dummyPopen` instance instead (see rambling above about need to revamp!).  That's not a context manager. I believe an addition as simple as this should let things proceed (untested):

```python
diff --git a/SCons/Action.py b/SCons/Action.py
index 67eeb1d86..d29a65c07 100644
--- a/SCons/Action.py
+++ b/SCons/Action.py
@@ -800,6 +800,8 @@ def _subproc(scons_env, cmd, error='ignore', **kw):
         # return a dummy Popen instance that only returns error
         class dummyPopen:
             def __init__(self, e): self.exception = e
+            def __enter__(self): return self
+            def __exit__(self, *args): pass
             def communicate(self, input=None): return ('', '')
             def wait(self): return -self.exception.errno
             stdin = None
```

Reply:
I supposed the two Java tets could also be adjusted to initialize only the tools it needs, as a separate/additional fix?  Not sure about Varables, which also seemed to fail.

Reply:
PR #4089 now includes DefaultEnvironment / Environment(tools=[limited list]) stanza for the two Java tests.

Reply:
I tested the enter/exit patch from above and it does eliminate the three `AttributeError` test failures.